### PR TITLE
JSONize move modes

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
SUMMARY: Infrastructure "JSONize move modes"

#### Purpose of change
Having things in JSON makes them easier to modify and expand on. 

#### Describe the solution
Add class `move_mode`, and two enums - `steed_type`, and `move_mode_type`.
`move_mode` contains all the data about a movement mode:

- The id
- The messages for changing to this mode for each steed type
- The messages for failing to change to this mode for each steed type
- Whether or not to stop hauling when this move mode is switched to
- The move mode to cycle to - e.g. walk to run, run to crouch, crouch to walk. This is dynmaically generated when data is loaded
- The type of the move mode - e.g. if you're sneaking, moving normally, or running
- The exertion level of the movement mode.
- The move speed multiplier of the mode
- The sound multiplier of the mode
- The stamina multiplier of the mode
- The color of the mode on panels and the character
- The letter of the mode in the move mode menu and the letter
- The name of the move mode

Add a file-global vector of move mode (id)s sorted by speed to `move_mode.cpp`, and a function to retrieve a const reference to it.
Use this to dynamically populate the move mode menu.

Use generic factory to load move modes, populating this vector in the finalize stage, and clearing it in the reset stage.
During this finalization stage, use the vector to populate the `cycle_to` values of the move modes.

Add functions to fetch the various information contained in move modes.

Add the JSON definitions for the move modes in `data/json/move_modes.json`.
Document these JSON definitions in `doc/MOVE_MODE.md`.

Modify `lang/extract_json_strings.py` to correctly translate the move mode messages.

Add functions `is_running()`, `is_walking()`, `is_crouching()`, `can_switch_to()` to Character.
Delete the `character_movemode` enum, and replace things checking this with either using the above function.
Convert functions taking it as an argument, pass a `move_mode_id` reference.
Change the `move_mode` variable in character to a `move_mode_id`.
Convert `can_run()` to be const, because there's no reason it shouldn't be.

Replace things determining some data based on move mode to returning the data from `move_mode` for that value.

Convert the various move mode changing functions to using move mode ids.
Change the move mode menu to dynamically populate based on the move modes loaded.

Convert achievements from using the enum to using move mode ids - this migrates smoothly, as the string ids are what the `enum_to_string` function for move modes previous returned.

Convert the vehicle follow speed to use the move mode speed multiplier, instead of different constant values for each move mode.

Convert tests to use move mode ids in place of the enum.

#### Testing
Spawn in and tame horse and mech
Switch to each movement mode these, and off. Set stamina to 0, and attempt to switch to running on each.
Ensure that the messages are correct.

Use the toggle crouch, toggle run, and cycle move mode, and ensure it puts you in the correct move mode.
Use the move mode menu, and see that the move modes are in order of fastest to slowest, and that choosing each mode puts you in the correct one, and that cycle works correctly.

Find a fence gate. Running into it does not open it, and walking into it does.

Load an existing save, and see that the movement statistics are correct.
Swim, and see that you move at different speeds in different move modes.
Attempt to ride a horse too close to enemies in run move mode, and be dismounted. Check that you are in walk move mode.
Mount a horse, go into run mode, then dismount. Ensure you are in walk mode.

Crouch behind a window, and see that you cannot see out of it.

Enter each move mode, then check the overmap. See that your character is white when walking, yellow when running, and light gray when crouching.

Use a tileset that shows move mode overlays (I used Ultica), and check that there's an overlay for crouch and run modes.

Run the stamina tests, and see that stamina burn is not impacted.

Run, and see that your move costs are half that of walking. Crouch, and see that your move costs are twice that of walking.

Find a pile of rubble, then verify that the various options for 'Dangerous terrain warning prompt' work correctly.

See that walking has 6 sound, running 9, and crouching 3 on normal terrain.

Running until out of stamina puts you into walk mode.

See that taking a selfie marks your character as sitting when crouching, and standing otherwise.

Crouching, walking, and running display C, W, and R, respectively in the panels, and are light blue, white, and red.

Horses pulling a wagon behind you do not run you over.

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
FYI: #38299

File: src/move_mode.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    mandatory( jo, was_loaded, "move_type", _type, make_flag_reader( move_types, "move type" ) );
+
+    optional( jo, was_loaded, "stamina_multiplier", _stamina_multiplier, 1.0 );
+    optional( jo, was_loaded, "sound_multiplier", _sound_multiplier, 1.0 );
+    optional( jo, was_loaded, "mech_power_use", _mech_power_use, 2 );
+    _mech_power_use *= -1;
+
+    optional( jo, was_loaded, "stop_hauling", _stop_hauling );
+};
```
Comment (ZhilkinSerg) [inline src/move_mode.cpp:72]:
```suggestion
}
```

File: data/json/move_modes.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  },
+  {
+    "type": "movement_mode",
+    "id": "crouch",
+    "character": "c",
+    "name": "crouch",
+    "exertion_level": "MODERATE_EXERCISE",
+    "change_good_none": "You start crouching.",
+    "change_good_animal": "You shlow your steed to a walk.",
+    "change_good_mech": "You reduce the power of your mech's leg servos t minimum.",
```
Comment (AMurkin) [inline data/json/move_modes.json:41]:
```suggestion
    "change_good_mech": "You reduce the power of your mech's leg servos to minimum.",
```

File: data/json/move_modes.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    "swim_speed_mod": -80
+  },
+  {
+    "type": "movement_mode",
+    "id": "crouch",
+    "character": "c",
+    "name": "crouch",
+    "exertion_level": "MODERATE_EXERCISE",
+    "change_good_none": "You start crouching.",
+    "change_good_animal": "You shlow your steed to a walk.",
```
Comment (AMurkin) [inline data/json/move_modes.json:40]:
```suggestion
    "change_good_animal": "You slow your steed to a walk.",
```

File: data/json/move_modes.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    "id": "run",
+    "character": "r",
+    "name": "run",
+    "exertion_level": "ACTIVE_EXERCISE",
+    "change_good_none": "You start running.",
+    "change_good_animal": "You spur your steed into a gallop.",
+    "change_good_mech": "You set the power of your mech's leg servos to maximum.",
+    "change_bad_none": "You're too tired to run.",
+    "change_bad_animal": "Your steed is too tired to go faster.",
+    "change_bad_mech": "Your steed is too tired to go faster.",
```
Comment (AMurkin) [inline data/json/move_modes.json:25]: Something about mech's leg servos here?

Comment (anothersimulacrum) [reply data/json/move_modes.json:25]: This is what it was previously - if you have any ideas, I'd be happy to change it.

Comment (AMurkin) [reply data/json/move_modes.json:27]: This option has not been implemented, AFAICT. But if he will be in the future, what about "Your mech's leg servos are unable to operate faster."?

**Comment by anothersimulacrum** [comment]
This should be ready to go.

**Pull request merged**